src/Interpreters/join_common.cpp

```diff
@@ -164,6 +177,83 @@ Columns materializeColumns(const Block & block, const Names & names)
     return materialized;
 }
 
+/// Join keys not casted in some cases:
+/// - for tables with `engine = Join` (in this case left_block hasn't rows)
```

Comment:
You can use a `Join` table as a righthand side of join, shouldn't we have normal left blocks in this case?

But I think our variant of type conversion doesn't apply to Join tables anyway -- the hash table is already built, but we convert the righthand values before building the hash table.